src/Graphics/Glyphs/GlyphGeom.cc

```diff
@@ -523,228 +567,632 @@ void GlyphGeom::generateSphere(const Point& center, double radius, int resolutio
+    Vector tempVec = eigvectors[0];
+    eigvectors[0] = eigvectors[1];
+    eigvectors[1] = tempVec;
+  }
+  if(eigvals[1] < eigvals[2])
+  {
+    double temp = eigvals[1];
+    eigvals[1] = eigvals[2];
+    eigvals[2] = temp;
+    Vector tempVec = eigvectors[1];
+    eigvectors[1] = eigvectors[2];
+    eigvectors[2] = tempVec;
+  }
+  if(eigvals[0] < eigvals[1])
+  {
+    double temp = eigvals[0];
+    eigvals[0] = eigvals[1];
+    eigvals[1] = temp;
+    Vector tempVec = eigvectors[0];
+    eigvectors[0] = eigvectors[1];
+    eigvectors[1] = tempVec;
+  }
 }
 
-void GlyphGeom::generateEllipsoid(const Point& center, Tensor& t, Vector &scaled_eigenvals, int resolution, const ColorRGB& color)
+void GlyphGeom::generateBox(const Point& center, Tensor& t, double scale, ColorRGB& node_color, bool normalize)
 {
-    Vector eig_vec1, eig_vec2, eig_vec3;
-    t.get_eigenvectors(eig_vec1, eig_vec2, eig_vec3);
+  double zeroThreshold = 0.000001;
+  double eigval1, eigval2, eigval3;
+  t.get_eigenvalues(eigval1, eigval2, eigval3);
+
+  Vector eigvals(abs(eigval1), abs(eigval2), abs(eigval3));
+  if(normalize)
+    eigvals.normalize();
+  eigvals *= scale;
+
+  std::vector<Vector> eigvectors(3);
+  t.get_eigenvectors(eigvectors[0], eigvectors[1], eigvectors[2]);
+
+  // Checks to see if eigenvalues are close to 0
+  bool eig_x_0 = eigvals.x() <= zeroThreshold;
+  bool eig_y_0 = eigvals.y() <= zeroThreshold;
+  bool eig_z_0 = eigvals.z() <= zeroThreshold;
+
+  // Set to 0 if below threshold
+  eigvals[0] = (!eig_x_0) * eigvals[0];
+  eigvals[1] = (!eig_y_0) * eigvals[1];
+  eigvals[2] = (!eig_z_0) * eigvals[2];
+
+  bool flatTensor = (eig_x_0 + eig_y_0 + eig_z_0) >= 1;
+  if(flatTensor)
+  {
+    reorderTensor(eigvectors, eigvals);
+
+    eig_x_0 = eigvals.x() <= zeroThreshold;
+    eig_y_0 = eigvals.y() <= zeroThreshold;
+    eig_z_0 = eigvals.z() <= zeroThreshold;
+    // Check for zero eigenvectors
+    if(eig_x_0)
+    {
+      eigvectors[0] = Cross(eigvectors[1], eigvectors[2]);
+    }
+    else if(eig_y_0)
+    {
+      eigvectors[1] = Cross(eigvectors[0], eigvectors[2]);
+    }
+    else if(eig_z_0)
+    {
+      eigvectors[2] = Cross(eigvectors[0], eigvectors[1]);
+    }
+  }
+
+  Transform rotate(Point(0.0, 0.0, 0.0), eigvectors[0], eigvectors[1], eigvectors[2]);
+  Transform trans = rotate;
+  trans.pre_translate((Vector) center);
+
+  // Rotate and translate points
+  Vector p1 = Vector(trans * Point(eigvals * Vector(-1.0, 1.0, 1.0)));
+  Vector p2 = Vector(trans * Point(eigvals * Vector(-1.0, 1.0, -1.0)));
+  Vector p3 = Vector(trans * Point(eigvals * Vector(1.0, 1.0, 1.0)));
+  Vector p4 = Vector(trans * Point(eigvals * Vector(1.0, 1.0, -1.0)));
+  Vector p5 = Vector(trans * Point(eigvals * Vector(-1.0, -1.0, 1.0)));
+  Vector p6 = Vector(trans * Point(eigvals * Vector(-1.0, -1.0, -1.0)));
+  Vector p7 = Vector(trans * Point(eigvals * Vector(1.0, -1.0, 1.0)));
+  Vector p8 = Vector(trans * Point(eigvals * Vector(1.0, -1.0, -1.0)));
```

Comment:
This is a good block to extract into a class/function. The 8 constant vectors then could be declared static constexpr in the new class, saving time and memory, and you'd reduce duplication. Calling code would look something like
```
BoxBuilder bb;
auto box = bb.makeSideVectors(trans, eigvals);
//....
//changed to 0-based
generateBoxSide(box[6], box[7], box[2], box[3], x_vec, node_color);
...
```

Reply:
I'll leave it to you to figure out what type the `auto` should be :)

Reply:
`Vector p5 = Vector(trans * Point(-eigvals.x(), -eigvals.y(), eigvals.z());`
What if I did this instead? This gets rid of instantiating the vector and doing the extra multiplication operation without needing a new class

Reply:
It's still a high degree of duplication

Reply:
Anytime there's a long list of similar-looking variables ending in 1,2,3,4,..., that is a smell and should be replaced with a function returning an array

Reply:
I'm using arrays for for both points and normals now. I refactored it so the points are generated in a loop because we previously had the permutations hard coded to get the 8 points. Is there any issues with this method?

`std::vector<Vector> points;
  for(int x = -1; x < 2; x+=2)
  {
    for(int y = -1; y < 2; y+=2)
    {
      for(int z = -1; z < 2; z+=2)
      {
        Point translated_point = trans * Point(x * eigvals.x(), y * eigvals.y(), z * eigvals.z());
        points.push_back(Vector(translated_point));
      }
    }
  }`

Reply:
A new-style better way to write each loop is:
```
for (int x : {-1, 1})
  for (int y : {-1, 1})
    for (int z : {-1, 1})
{
 ...
}
```

Reply:
And I'd still at the very least extract this into a separate function.

Reply:
And you can make the body a one-liner:
```
        points.emplace_back(trans * Point(x * eigvals.x(), y * eigvals.y(), z * eigvals.z()));
```

Reply:
I like that loop style much better. Why should we separate point generation into another function? I don't see any place we could reuse that code. Every other glyph has their points being generated in the same function and boxes have a much simpler implementation.

Reply:
I put it into a separate function. I decided against separating point generation for other glyphs because we can't without duplicating the nested for loop. It's too intertwined with generating normals unlike boxes. Plus we would have to pass along tons of parameters.